File: CPP/ByTier/C/1605_C_Dominant_Character.cpp
// Problem: CF 1605 C - Dominant Character
// https://codeforces.com/contest/1605/problem/C

/*
 * Problem: Find the shortest substring where 'a' occurs strictly more times than both 'b' and 'c'
 * Algorithm: Brute force approach with pre-defined patterns
 * Time Complexity: O(n) where n is the length of string s, as we check at most 7 predefined substrings
 * Space Complexity: O(1) as we use only a constant amount of extra space
 * 
 * The approach is based on observing that the minimum valid substring must be one of these 7 patterns:
 * "aa", "aba", "aca", "abca", "acba", "abbacca", "accabba"
 * These patterns cover all possible minimal configurations where 'a' > 'b' and 'a' > 'c'
 * We check if any of these patterns exist in the input string and return the length of the first match
 */

#include <iostream>
#include <string>

using namespace std;

int main() {
  int t;
  cin >> t;
  while (t--) {
    int n, i;
    cin >> n;
    string s,
        ans[7] = {"aa", "aba", "aca", "abca", "acba", "abbacca", "accabba"};
    cin >> s;
    int res = -1;
    for (int i = 0; i < 7; i++) {
      // Check if current pattern exists in the string
      if (s.find(ans[i]) != -1) {
        res = ans[i].length();  // Store length of first matching pattern
        break;  // Stop as we want the smallest length
      }
    }
    cout << res << endl;
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/C/1605_C_Dominant_Character.py
# Problem: CF 1605 C - Dominant Character
# https://codeforces.com/contest/1605/problem/C

"""
Problem: Find the length of the smallest substring where 'a' occurs strictly more times than 'b' and 'c', and the substring length is at least 2.

Approach: 
- Since we need the smallest valid substring, we can check for all possible small patterns
- The key insight is that we only need to check substrings of length up to 9 (based on the provided patterns)
- We look for specific patterns that are known to be valid candidates
- Time Complexity: O(n) where n is the length of string s
- Space Complexity: O(1) as we use constant extra space
"""

for _ in range(int(input())):
    n = int(input())
    s = input()
    # Check for predefined patterns that satisfy the conditions
    for i in ["aa", "aba", "aca", "abca", "acba", "abbacca", "accabba"]:
        if i in s:  # If pattern exists in string
            print(len(i))  # Print length of first matching pattern
            break  # Exit loop after first match
    else:
        print(-1)  # If no pattern found, print -1


# https://github.com/VaHiX/CodeForces/